### ISSUE TITLE
RFC: Add command line option to lxd to disable vm support

### ISSUE THREAD

**Issue by geaaru** [lxc/lxd#10905]
# Required information

 * Distribution: all
 * Distribution version: all
 * LXD version: >=5.5

# Issue description

There are different use cases, for example using LXD in LXD in a restricted container, etc. where LXD could be used for containers and it isn't possible using the VM driver because the /dev/vsock kernel driver is not available.

With the upgrade to LXD 5.5 seems mandatory the activation of the listening socker for /dev/vsock also when is not used.

Speaking on this in the IRC channel the suggestion is avoid the loading of `kvm` kernel driver but it seems not too correct. From a user aspect I think that could be ok to have kvm kernel driver for other needs, like to have the qemu binary but simply don't use VM support on LXD.

Could be a good idea to have a `--disable-qemu-driver` option to lxd binary?

I dunno if there is another way to disable qemu driver.
Thanks in advance

**Comment by tomponline** [comment]
If `/dev/vsock` isn't available in the container, what is happening that is causing a problem? 

Having /dev/vsock available shouldn't be a requirement for LXD to start up. Is that the issue?

**Comment by geaaru** [comment]
hi @tomponline if the `/dev/vsock` LXD doesn't start.
```
DEBUG  [2022-09-08T17:16:50Z] Initializing database gateway                
INFO   [2022-09-08T17:16:50Z] Starting database node                        id=1 local=1 role=voter
INFO   [2022-09-08T17:16:52Z] Loading daemon configuration                 
INFO   [2022-09-08T17:16:52Z] Closing socket                                socket=/var/lib/lxd/unix.socket type="REST API Unix socket"
INFO   [2022-09-08T17:16:52Z] Closing socket                                socket=/var/lib/lxd/devlxd/sock type="devlxd socket"
ERROR  [2022-09-08T17:16:52Z] Failed to start the daemon                    err="listen vsock: open /dev/vsock: no such file or directory"
INFO   [2022-09-08T17:16:52Z] Starting shutdown sequence                    signal=interrupt
DEBUG  [2022-09-08T17:16:52Z] Cancel ongoing or future gRPC connection attempts 
INFO   [2022-09-08T17:16:52Z] Stop database gateway                        
INFO   [2022-09-08T17:16:52Z] Not unmounting temporary filesystems (instances are still running) 
INFO   [2022-09-08T17:16:52Z] Daemon stopped 
```
So, if this something wrong could be an issue too but in my case the RFC is for ask an option that permit to disable the instance type VM instead of just trying to detect it from existing rootfs.
I had few time to read the code about DriverStatuses() that report the list of the driver supported and it isn't so clear to me how it works that LXD detect system in this case.

**Comment by tomponline** [comment]
Inside the container is `/dev/kvm` available?

**Comment by geaaru** [comment]
Yes.
```
# ls /dev
console  core  fd  full  fuse  kvm  lxd  mapper  mqueue  net  null  ptmx  pts  random  shm  stderr  stdin  stdout  tty  urandom  vhost-net  vhost-scsi  vhost-vsock  zero
```

**Comment by tomponline** [comment]
OK do you know why its being passed through into the container? I suspect the issue will go away if you don't indicate to the container the kvm is available by passing /dev/kvm into it.

**Comment by tomponline** [comment]
> From a user aspect I think that could be ok to have kvm kernel driver for other needs, like to have the qemu binary but simply don't use VM support on LXD.

I see that you're intending to use qemu (which is why /dev/kvm is required) but not LXD VMs, so not providing /dev/vsock.

Can you confirm though that LXD will start if no /dev/kvm available, just for completeness.

**Comment by geaaru** [comment]
Tested with a cloned container. Yeah, I confirm that disable KVM permits LXD bootstrap:
```
WARNING[2022-09-09T14:36:29Z] Instance type not operational                 driver=qemu err="KVM support is missing (no /dev/kvm)" type=virtual-machine
```

**Comment by stgraber** [comment]
That'd be a bug, LXD should be able to start just fine in such scenarios and simply not support running VMs.
Looks like we need to fix the vsock listener.

**Comment by tomponline** [comment]
@stgraber yeah we should be able to fix this by adding detection of /dev/vsock to the VM supported logic, which would then prevent this listener from being tried.

**Comment by geaaru** [comment]
IMHO, i think that could be good also to have a way to disable vm support when both vsock and kvm devices are available.

**Comment by stgraber** [comment]
If you need to disable VMs, the best is to set a VM limit of 0 on your project.

**Comment by geaaru** [comment]
Yes, correct. But this doesn't avoid to open a listening socket with vsock kernel.
I have really few experience with vsock kernel driver. Sharing the same /dev/vsock between multiple containers could be a security hole if multiple users starts LXD in container and using the same /dev/vsock?
thanks in advance for any suggestions.

**Comment by tomponline** [comment]
If fix the bug so that if /dev/vsock isn't provided to the container it will be considered as not supporting VMs (even if /dev/kvm is passed in), then LXD will start as normal but will not allow VMs to be used.

**Comment by geaaru** [comment]
Personally, i think that to have an option to force shutdown is also good (because you have the same problem is there are in the container both /dev/kvm and /dev/vsock devices but you don't want VM supporto) but i leave to you how proceed ;)
Thanks for the very fast feedback and support.

**Comment by Marietto2008** [comment]
I have the same issue :

https://github.com/lxc/lxd/issues/11597#issuecomment-1518589443
